platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1573,58 +1573,47 @@ private void assertUsageIsMutable() {
+     * method will not emit a warning. We will want to remove this condition once the role-locked configuration creation
+     * APIs are de-incubated.</p>
+     *
+     * <p>If a usage is being disabled for a detached configuration, this method does not emit a warning.
+     * KMP disables consumable for detached configurations even though this is not necessary. This condition
+     * can be disabled with an internal feature flag system property for testing purposes.</p>
+     *
+     * <p>This method will also not warn if the current and new value are the same. We also want to eventually remove
+     * this condition, as build logic should not attempt to mutate a configuration's roles if that configuration
+     * has a role set upon creation. This condition can be disabled with an internal feature flag system property
+     * for testing purposes.</p>
      *
      * The eventual goal is that all configuration usage be specified upon creation and immutable thereafter.
      */
-    private void maybeWarnOnChangingUsage(String usage, boolean current) {
-        if (isDetachedConfiguration() || isInLegacyRole()) {
+    private void maybeWarnOnChangingUsage(String methodName, boolean current, boolean newValue) {
+        if (isInLegacyRole()) {
             return;
         }
 
-        String msgTemplate = "Allowed usage is changing for %s, %s. Ideally, usage should be fixed upon creation.";
-        String changingUsage = usage + " was " + !current + " and is now " + current;
+        // KGP continues to set the already-set value for a given usage even though it is already set
+        boolean redundantChange = current == newValue;
 
-        DeprecationLogger.deprecateBehaviour(String.format(msgTemplate, getDisplayName(), changingUsage))
-            .withAdvice("Usage should be fixed upon creation.")
-            .willBeRemovedInGradle9()
-            .withUpgradeGuideSection(8, "configurations_allowed_usage")
-            .nagUser();
-    }
+        // KGP disables `consumable` on detached configurations even though this is not necessary
+        boolean disableUsageForDetached = isDetachedConfiguration() && !newValue;
 
-    private void maybeWarnOnRedundantUsageActivation(String usage, String method) {
-        if (!isSpecialCaseOfRedundantUsageActivation()) {
-            String msgTemplate = "The %s usage is already allowed on %s.";
-            DeprecationLogger.deprecateBehaviour(String.format(msgTemplate, usage, getDisplayName()))
-                .withAdvice(String.format("Remove the call to %s, it has no effect.", method))
-                .willBeRemovedInGradle9()
-                .withUpgradeGuideSection(8, "redundant_configuration_usage_activation")
-                .nagUser();
+        // This property exists to allow KGP to test whether they have properly resolved this deprecation.
+        // This property WILL be removed without warning.
```

Comment:
Sure, but you know others will find it.  Definitely a broader concern than just this PR though.

Reply:
The options I see are:
1. Remove the flag entirely. This makes it hard for KMP to verify they've fixed their code
2. Implement preliminary deprecations holistically. I don't see this happening right now
3. Make it a full deprecation. This would lead to a LOT of warnings. I know we just got a little softer on deprecations, but this would be a ton of spam and I think we should avoid that. 
4. Leave the flag in, tell KMP, they fix the problem, we remove the flag. 

As far as people seeing this code, adding the flag to their build, and committing it to source control... I think that's on them. The worst thing that happens is they have a flag that does nothing in their gradle.properties once we remove it. There is a clear warning that this will be removed and the flag is marked `internal`.

Reply:
Another options is to just accept accept that redundant mutations on role-locked configurations are okay. I'm not saying we _should_ do this, but saying `canBeResolved = false` on an already non-resolvable configuration _can_ be handled gracefully without failing the build (by doing nothing). 

I agree its ugly and we should probably do something about it, but there are paths forward here where we just do nothing and let people have ugly build logic. 

In any case, once role-lockec configurations become a thing, we would be removing the legacy configuration and would then remove these methods as well. 

We might have more success and annoy users less if we don't warn for redundant usage calls _ever_ and just deprecate these methods when the time comes